File: src/nostr/nip_04_test.rs
use super::*;

    #[test]
    fn test_nip04_roundtrip() {
        let shared_secret = [42u8; 32];
        let plaintext = "Hello, Nostr!";
        
        let encrypted = encrypt_nip04(&shared_secret, plaintext).unwrap();
        assert!(encrypted.contains("?iv="));
        
        let decrypted = decrypt_nip04(&shared_secret, &encrypted).unwrap();
        assert_eq!(plaintext, decrypted);
    }

    #[test]
    fn test_nip04_invalid_format() {
        let shared_secret = [42u8; 32];
        let result = decrypt_nip04(&shared_secret, "not-base64-content");
        assert!(result.is_err());
        assert_eq!(result.unwrap_err().to_string(), "error: Invalid NIP-04 format");
    }

    #[test]
    fn test_nip04_wrong_key() {
        let shared_secret1 = [1u8; 32];
        let shared_secret2 = [2u8; 32];
        let plaintext = "Sensitive data";
        
        let encrypted = encrypt_nip04(&shared_secret1, plaintext).unwrap();
        let result = decrypt_nip04(&shared_secret2, &encrypted);
        
        // Either it fails to unpad or returns garbage
        assert!(result.is_err());
    }

    #[test]
    fn test_nip04_tampered_ciphertext() {
        let shared_secret = [42u8; 32];
        let plaintext = "Another secret";
        
        let encrypted = encrypt_nip04(&shared_secret, plaintext).unwrap();
        // Tamper by removing the last character of the ciphertext to break base64/padding
        let parts: Vec<&str> = encrypted.split("?iv=").collect();
        let tampered_ct = &parts[0][..parts[0].len()-1];
        let tampered_encrypted = format!("{}?iv={}", tampered_ct, parts[1]);
        
        let result = decrypt_nip04(&shared_secret, &tampered_encrypted);
        // This will now consistently fail due to malformed base64 or invalid padding
        assert!(result.is_err());
    }

File: src/nostr/nip04.test.ts
import { describe, it, expect } from 'vitest';
import { encryptNip04, decryptNip04 } from './nip04.ts';

const SHARED_SECRET = new Uint8Array(32).fill(42);

describe('NIP-04', () => {
  it('roundtrip', async () => {
    const plaintext = 'Hello, Nostr!';
    const encrypted = await encryptNip04(SHARED_SECRET, plaintext);
    expect(encrypted).toContain('?iv=');
    const decrypted = await decryptNip04(SHARED_SECRET, encrypted);
    expect(decrypted).toBe(plaintext);
  });

  it('rejects invalid format', async () => {
    await expect(decryptNip04(SHARED_SECRET, 'not-base64-content')).rejects.toThrow('Invalid NIP-04 format');
  });

  it('wrong key fails to decrypt', async () => {
    const key1 = new Uint8Array(32).fill(1);
    const key2 = new Uint8Array(32).fill(2);
    const encrypted = await encryptNip04(key1, 'Sensitive data');
    await expect(decryptNip04(key2, encrypted)).rejects.toThrow();
  });

  it('tampered ciphertext fails', async () => {
    const encrypted = await encryptNip04(SHARED_SECRET, 'Another secret');
    const parts = encrypted.split('?iv=');
    const tamperedCt = parts[0].slice(0, -1);
    const tamperedEncrypted = `${tamperedCt}?iv=${parts[1]}`;
    await expect(decryptNip04(SHARED_SECRET, tamperedEncrypted)).rejects.toThrow();
  });
});
